### PR TITLE
Bug fix for del() with multiple Foreign Key aliases to the same Primary. 

### LINKED ISSUE #379: bug with del()
When attempting to delete I get the following error:

>> del(EXP2.Session & 'subject_id=481102')
Error using  ~ 
Too many input arguments.

Error in dj.Relvar/del (line 95)
                            if ~fks.aliased

'subject_id' is the primary key for table EXP2.Session, and this always worked before. I noticed it after upgrading to DataJoint version 3.4.3. 
del() with few other tables that are under EXP2.Session with the same key that I tested worked fine

MATLAB 2018b
DataJoint version 3.4.3

bug

### PR DESCRIPTION
**Pull request by zitrosolrac**
Fix #379

### CODE REVIEW

**Review by guzman-raphael — APPROVED** [review]
@zitrosolrac @A-Baji Great work here guys! I know some users who will be happy to see this patched. :smile:

File: +dj/Relvar.m [hunk 1]
```diff
@@ -92,23 +92,22 @@ function cleanup(self)
                                 @(x) strcmp(x.from, rels(ix).fullTableName), ...
                                 self.schema.conn.foreignKeys, 'uni', true);
                             fks = self.schema.conn.foreignKeys(fk_index);
-                            if ~fks.aliased
+                            if ~all(fks.aliased)
```
Comment (zfj1) [inline +dj/Relvar.m:95]: This line still doesn't work for us. Should read `if ~all([fks.aliased])`

**Comment by zfj1** [comment]
@guzman-raphael Requesting a further fix to get this working

**Comment by kabilar** [comment]
Thanks for the report @zfj1.  We will review and get back to you.

File: +dj/Relvar.m (hunk 1 above)
Comment (A-Baji) [reply +dj/Relvar.m:95]: Hey @zfj1, could you elaborate on why this change was necessary on your end? We would like to recreate the issue.

Comment (zfj1) [reply +dj/Relvar.m:95]: If a table has multiple foreign keys, then `fks` will be a `1xN` struct array, e.g.
```
fks = 

  1×3 struct array with fields:

    attrs
    ref
    ref_attrs
    from
    primary
    multi
    aliased
```
In this case, `fks.aliased` will return multiple outputs rather than an array of outputs, e.g.,
```
fks.aliased

ans =

  logical

   0


ans =

  logical

   0


ans =

  logical

   0
```
In this case, `all(fks.aliased)` will return an `Too many input arguments` error, since you are technically calling `all(false,false,false)`. By calling `all([fks.aliased])` MATLAB interprets the call as `all([false,false,false])` which will be handled correctly and return `false`.

Comment (A-Baji) [reply +dj/Relvar.m:95]: @zfj1 , The test case and implementation has been included in https://github.com/datajoint/datajoint-matlab/pull/395.

Turns out `all(fks.aliased)` works with only two arguments, but not three or more, which we weren't testing for. Thanks for catching this.

**Comment by guzman-raphael** [comment]
@zfj1 Thanks for the report. :handshake: 

We've addressed this now in our new release `3.5.0`. Please let us know if you run into any further issues.

**Comment by zfj1** [comment]
Hi @guzman-raphael @kabilar,

We've run into more issues here. The offending table:

```
:: BreedingPair
-> AnimalSource #gives an id number to this breeding pair
---
(male_id) -> Animal
(female_id) -> Animal
```

This table has 2/3 aliased foreign keys (so `all([fks.aliased])` is `false`), but the restriction fails. Specifically, it seems to always try to delete every entry from this table when deleting an animal from the `Animal` table, even when the given animal is not a member of a `BreedingPair`. I've implemented a fix that works for this particular table [on my branch](https://github.com/zfj1/datajoint-matlab/blob/45f77cf699765a3dfefe0411402c639698b4d58f/+dj/Relvar.m) which you can take a look at.

Basically, one needs to check if `any()` foreign keys are aliased, not `all()`. But the foreign keys should also be cross-referenced with the referee, thus `~any([fks.aliased] & fks_index_i)` is the appropriate check. This change requires a change to the else block of the condition to parse properly.

I would strongly suggest testing this on a more pathological table, with several aliased and unaliased FKs, and FKs to tables with multiple primary keys. Accidentally deleting an entire table due to a failed restriction is certainly something we're keen to avoid :).

**Comment by kabilar** [comment]
Thanks for the report @zfj1.  We will review and provide you with an estimated timeline for the fix.

**Comment by guzman-raphael** [comment]
Thanks @zfj1 for the detailed report! :handshake: This is very similar to a [test case](https://github.com/datajoint/datajoint-matlab/blob/42804e9c4cfd85e16bf3be7857c545dcfa25a61b/tests/TestDelete.m#L39-L66) we already have that passes, however, without the FK on the primary. 

A few questions come to mind:
1. Is the definition you've provided exactly how you are currently defining it? I would expect some error in ambiguity related to the renaming.
1. Have you by chance ran the tests to see if your fix passes our entire test suite?

In an initial review, nothing obvious comes to mind yet but I'd generally approach this by:

1. Creating a new test case that reproduces per [above](https://github.com/datajoint/datajoint-matlab/pull/386#issuecomment-1194889391) resulting in a failure case
1. Apply the fix
1. Iterate until the entire suite of tests pass

@zitrosolrac Would you help starting to investigate this?

We'll have a better idea on the estimated time to solution once we understand better why the prior logic did not account for this case.

**Comment by zitrosolrac** [comment]
@zfj1 @guzman-raphael As an update to your inquiry, I can relay that you're suggested changes were implemented into relvar.m and all of the tests in our test suite passed. Thank you for you contribution and we will consider merging this into the app soon!
